juju/bundle.py

```diff
@@ -156,311 +171,663 @@ def resolve(self, reference):
                 reference = ref
         return reference
 
-    async def addCharm(self, change_id, charm, series):
+
+def get_charm_series(path):
+    """Inspects the charm directory at ``path`` and returns a default
+    series from its metadata.yaml (the first item in the 'series' list).
+
+    Returns None if no series can be determined.
+    """
+    md = Path(path) / "metadata.yaml"
+    if not md.exists():
+        return None
+    data = yaml.load(md.open())
+    series = data.get('series')
+    return series[0] if series else None
+
+
+class ChangeSet:
+
+    def __init__(self, changes):
+        self.changes = changes
+
+    # sorted does a topological sort of the changes
+    def sorted(self):
+        if len(self.changes) == 0:
+            return []
+
+        changes = {}
+        for change in self.changes:
+            changes[change.id_] = set(change.requires)
+        sorted_changes = toposort_flatten(changes)
+        results = []
+        for change_id in sorted_changes:
+            for change in self.changes:
+                if change_id == change.id_:
+                    results.append(change)
+                    break
+        return results
+
+
+class ChangeInfo:
+    _toPy = {}
+
+    def __init__(self, change_id, requires):
+        self.change_id = change_id
+        self.requires = requires
+
+    @classmethod
+    def from_dict(cls, self, data):
+        """
+        from_dict converts a data bag into fields on a class instance.
+        If a value is missing from the data, then None is assigned to the field
+        instance value.
+        """
+        d = (data or {})
+        for k, v in cls._toPy.items():
+            if k in d:
+                setattr(self, v, d[k])
+            else:
+                setattr(self, v, None)
+
+
+class AddApplicationChange(ChangeInfo):
+    _toPy = {'charm': 'charm',
+             'series': 'series',
+             'application': 'application',
+             'options': 'options',
+             'constraints': 'constraints',
+             'storage': 'storage',
+             'devices': 'devices',
+             'endpoint-bindings': 'endpoint_bindings',
+             'resources': 'resources',
+             'num-units': 'num_units'}
```

Comment:
is this just here to map - to _ ? Can we just do that transition automatically in place? e.g. xxx.replace('-', '_') or xxx.standardize() or something?